File: keycard_wallet/tests/force_unpower.py
#!/usr/bin/env python3
"""
Forces the card in the first available reader into the unpowered state via
PC/SC SCARD_UNPOWER_CARD. Run immediately before a wallet command to simulate
the power-loss condition reported on some USB reader/driver combinations.

Either:
- pcscd re-powers the card on the next SCardConnect, so wallet
commands will succeed without triggering the retry path.
- the card stays unpowered, triggering TransportError
and exercising the retry wrapper in pair() / setup_communication_with_pairing().
"""
import sys
from smartcard.scard import (
    SCardEstablishContext, SCardListReaders, SCardConnect, SCardDisconnect,
    SCARD_SCOPE_USER, SCARD_SHARE_SHARED,
    SCARD_PROTOCOL_T0, SCARD_PROTOCOL_T1,
    SCARD_UNPOWER_CARD,
)

hresult, hcontext = SCardEstablishContext(SCARD_SCOPE_USER)
hresult, reader_list = SCardListReaders(hcontext, [])

if not reader_list:
    print("force_unpower: no readers found, skipping.")
    sys.exit(0)

hresult, hcard, _ = SCardConnect(
    hcontext,
    reader_list[0],
    SCARD_SHARE_SHARED,
    SCARD_PROTOCOL_T0 | SCARD_PROTOCOL_T1,
)

if hresult != 0:
    print(f"force_unpower: SCardConnect failed (hresult={hresult:#010x}), skipping.")
    sys.exit(0)

SCardDisconnect(hcard, SCARD_UNPOWER_CARD)
print("force_unpower: card powered down.")

File: lez/keycard_wallet/src/bin/force_unpower.rs
#![expect(
    clippy::print_stdout,
    reason = "This is a CLI test helper, printing to stdout is expected and convenient"
)]

//! Forces the card in the first available reader into the unpowered state via PC/SC
//! `SCARD_UNPOWER_CARD`. Run immediately before a wallet command to simulate the power-loss
//! condition reported on some USB reader/driver combinations.
//!
//! Either:
//! - pcscd re-powers the card on the next `SCardConnect`, so wallet commands will succeed without
//!   triggering the retry path.
//! - the card stays unpowered, triggering a PC/SC transport error (`keycard_rs::Error::Io`) and
//!   exercising the reconnect-and-retry wrapper in `KeycardWallet::connect()`.

fn main() {
    let context = match pcsc::Context::establish(pcsc::Scope::User) {
        Ok(context) => context,
        Err(e) => {
            println!("force_unpower: failed to establish PC/SC context ({e}), skipping.");
            return;
        }
    };

    let readers = match context.list_readers_owned() {
        Ok(readers) => readers,
        Err(e) => {
            println!("force_unpower: failed to list readers ({e}), skipping.");
            return;
        }
    };

    let Some(reader) = readers.first() else {
        println!("force_unpower: no readers found, skipping.");
        return;
    };

    let card = match context.connect(reader, pcsc::ShareMode::Shared, pcsc::Protocols::ANY) {
        Ok(card) => card,
        Err(e) => {
            println!("force_unpower: connect failed ({e}), skipping.");
            return;
        }
    };

    if let Err((_card, e)) = card.disconnect(pcsc::Disposition::UnpowerCard) {
        println!("force_unpower: disconnect failed ({e}), skipping.");
        return;
    }

    println!("force_unpower: card powered down.");
}
